### PR TITLE
Handle 401s from backend proxy upgrades.

### PR DESCRIPTION
**Pull request by ncdc**

### CODE REVIEW

File: pkg/util/httpproxy/upgradeawareproxy.go [hunk 1]
```diff
@@ -172,6 +173,29 @@ func (p *UpgradeAwareSingleHostReverseProxy) serveUpgrade(w http.ResponseWriter,
...
+	if err != nil {
+		glog.Errorf("Error writing request to backend: %s", err)
+		return
+	}
+
+	resp, err := http.ReadResponse(bufio.NewReader(backendConn), req)
+	if err != nil {
+		glog.Errorf("Error reading response from backend: %s", err)
+		w.Write([]byte(err.Error()))
+		w.WriteHeader(http.StatusInternalServerError)
```
Comment (liggitt) [inline pkg/util/httpproxy/upgradeawareproxy.go]: header comes first, right?

Comment (ncdc) [inline pkg/util/httpproxy/upgradeawareproxy.go]: Yeah, my bad

Comment (liggitt) [inline pkg/util/httpproxy/upgradeawareproxy.go]: yeah, write will force writeheader(ok)

File: pkg/util/httpproxy/upgradeawareproxy.go [hunk 2]
```diff
@@ -172,6 +173,29 @@ func (p *UpgradeAwareSingleHostReverseProxy) serveUpgrade(w http.ResponseWriter,
...
+		glog.Errorf("Error reading response from backend: %s", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte(err.Error()))
+		return
+	}
+
+	if resp.StatusCode == http.StatusUnauthorized {
+		glog.Errorf("Got unauthorized error from backend for: %s %s", req.Method, req.URL)
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte(err.Error()))
```
Comment (liggitt) [inline pkg/util/httpproxy/upgradeawareproxy.go]: Write "Internal Server Error", not `err`

File: pkg/util/httpproxy/upgradeawareproxy.go [hunk 3]
```diff
@@ -182,17 +206,12 @@ func (p *UpgradeAwareSingleHostReverseProxy) serveUpgrade(w http.ResponseWriter,
 	// NOTE: from this point forward, we own the connection and we can't use
 	// w.Header(), w.Write(), or w.WriteHeader any more
 
-	addAuthHeaders(req, p.clientConfig)
-
-	err = req.Write(backendConn)
+	err = resp.Write(requestHijackedConn)
```
Comment (liggitt) [inline pkg/util/httpproxy/upgradeawareproxy.go]: call `removeChallengeHeaders` on the resp first

**Pull request merged**